### ISSUE TITLE
Anansi does not respect the control rate

### ISSUE THREAD

**Issue by cgueret** [bbcarchdev/anansi#63]
When a URI is added to the queue Anansi does not wait 1s between fetching another URI from the same root. The query https://github.com/bbcarchdev/anansi/blob/develop/crawler/queues/db.c#L554-L563 may need to be changed to take the rate in account in the value for earliest update. Or https://github.com/bbcarchdev/anansi/blob/develop/crawler/queues/db.c#L1031 should take the rate into account when inserting a new resource.

**Comment by townxelliot** [comment]
To analyse the rate at which Anansi will crawl a site, you can follow these steps:
1.  Download this file: 
   [shakespeare.txt](https://github.com/bbcarchdev/anansi/files/583630/shakespeare.txt)
2.  Put that file into a directory, then start Python 3's simple HTTP server in that directory on port 3000:
   
   ```
   mkdir ~/server
   cd ~/server
   cp shakespeare.txt ~/server/shakespeare.ttl
   python3 -m http.server 3000
   ```
3.  Start an Anansi instance; the simplest way is via the Acropolis docker-compose.yml, which starts a single Anansi with one thread.
4.  On the machine where you're running Anansi, alias rdfserver.local to the IP address of the machine where the Python web server is running (e.g. by an entry to /etc/hosts).
   
   If you're using Acropolis docker-compose, you can do:
   
   `docker exec acropolis_anansi_1 /bin/bash -c "echo -e \"172.17.0.1  rdfserver.local\" >> /etc/hosts"`
   
   where 172.17.0.1 is the IP address of the docker0 interface on the Python webserver machine.
5.  Also add a similar alias to the /etc/hosts file of the Python web server's machine (this may be necessary for routing from the docker0 interface to the IP address of the machine running the web server):
   
   `127.0.0.1  rdfserver.local`
6.  Seed Anansi with the start URI, http://rdfserver.local:3000/shakespeare.ttl, which will make it fetch the file from your Python webserver.
   
   If you're using Acropolis docker-compose, you can do this with:
   
   `docker exec acropolis_anansi_1 /usr/bin/crawler-add -f http://rdfserver.local:3000/shakespeare.ttl`

Anansi should now start crawling the Python webserver site, starting at shakespeare.ttl, which will return 404s for every URI except the first one.

The important part is the gap between the requests: even a single-threaded Anansi will make multiple requests to the same domain within a second, if the server responds fast enough.

Here's an example from my log:

```
172.19.0.8 - - [10/Nov/2016 14:49:40] "GET /Person/shakespeare-characters/id/DukeofGloucester-h62 HTTP/1.1" 404 -
172.19.0.8 - - [10/Nov/2016 14:49:40] code 404, message File not found
172.19.0.8 - - [10/Nov/2016 14:49:40] "GET /Person/shakespeare-characters/id/DukeofGloucester-r3 HTTP/1.1" 404 -
172.19.0.8 - - [10/Nov/2016 14:49:40] code 404, message File not found
172.19.0.8 - - [10/Nov/2016 14:49:40] "GET /Person/shakespeare-characters/id/DukeofMilan-tg HTTP/1.1" 404 -
172.19.0.8 - - [10/Nov/2016 14:49:40] code 404, message File not found
172.19.0.8 - - [10/Nov/2016 14:49:40] "GET /Person/shakespeare-characters/id/DukeofNorfolk-h63 HTTP/1.1" 404 -
172.19.0.8 - - [10/Nov/2016 14:49:40] code 404, message File not found
172.19.0.8 - - [10/Nov/2016 14:49:40] "GET /Person/shakespeare-characters/id/DukeofNorfolk-h8 HTTP/1.1" 404 -
172.19.0.8 - - [10/Nov/2016 14:49:40] code 404, message File not found
172.19.0.8 - - [10/Nov/2016 14:49:40] "GET /Person/shakespeare-characters/id/DukeofNorfolk-r3 HTTP/1.1" 404 -
```

The server is responding very quickly with 404s, which means it can serve all the requests from Anansi; but even if this is the case, Anansi shouldn't attempt to fetch more than one URI per second from a single domain, which it is doing here (6 GET requests within the same second).

**Comment by nevali** [comment]
Okay, the question is why https://github.com/bbcarchdev/anansi/blob/develop/crawler/queues/db.c#L561 isn't being respected — why is `root.earliest_update` not being set appropriately?

I've not checked completely, but I've a feeling that an `UPDATE` should be added at https://github.com/bbcarchdev/anansi/blob/develop/crawler/queues/db.c#L1116 to reset `earliest_update` when the root is already present.

**Comment by nevali** [comment]
duplicate of #41

**Comment by nevali** [comment]
Resolved in e13639d
